src/crypto/mod.rs

```diff
@@ -156,47 +171,129 @@ impl fmt::Debug for SignatureShare {
     }
 }
 
-/// A secret key.
-#[derive(Clone, PartialEq, Eq, Rand)]
-pub struct SecretKey(Fr);
+/// A secret key; wraps a single prime field element. The field element is
+/// heap allocated to avoid any stack copying that results when passing
+/// `SecretKey`'s around as arguments or as return values.
+#[derive(PartialEq, Eq)]
+pub struct SecretKey(Box<Fr>);
 
-impl fmt::Debug for SecretKey {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let uncomp = self.public_key().0.into_affine().into_uncompressed();
-        let bytes = uncomp.as_ref();
-        write!(f, "SecretKey({:?})", HexBytes(bytes))
+impl Default for SecretKey {
+    fn default() -> Self {
+        let mut fr = Fr::zero();
+        match SecretKey::from_mut_ptr(&mut fr as *mut Fr) {
+            Ok(sk) => sk,
+            Err(e) => panic!("{}", e),
+        }
     }
 }
 
-impl Default for SecretKey {
-    fn default() -> Self {
-        SecretKey(Fr::zero())
+impl Rand for SecretKey {
+    fn rand<R: Rng>(rng: &mut R) -> Self {
+        let mut fr = Fr::rand(rng);
+        match SecretKey::from_mut_ptr(&mut fr as *mut Fr) {
+            Ok(sk) => sk,
+            Err(e) => panic!("{}", e),
+        }
+    }
+}
+
+impl Clone for SecretKey {
+    fn clone(&self) -> Self {
+        let mut fr = *self.0;
+        match SecretKey::from_mut_ptr(&mut fr as *mut Fr) {
+            Ok(sk) => sk,
+            Err(e) => panic!("{}", e),
+        }
     }
 }
 
 impl Drop for SecretKey {
     fn drop(&mut self) {
-        let ptr = self as *mut Self;
+        self.zero_secret_memory();
+        if let Err(e) = self.munlock_secret_memory() {
+            panic!("{}", e);
```

Comment:
Would it make sense to only `print!` if `thread::panicking`? Otherwise we'd always get a "panic while panicking" if munlock fails on two or more instances.